torch/nn/parallel/distributed.py

```diff
@@ -676,6 +676,9 @@ def no_sync(self):
             self.require_backward_grad_sync = old_require_backward_grad_sync
 
     def forward(self, *inputs, **kwargs):
+        if torch.is_grad_enabled() and self.require_backward_grad_sync:
+            self.logger.set_runtime_stats()
```

Comment:
Now that we have sampling I wonder if it's more performant to check the sampling here and only call into C++ if needed, might be faster than dropping/reacquiring GIL

Reply:
in this case, we need to expose should_collect_runtime_stats() to python